text/0008-ingest-gem-content.md

```diff
@@ -0,0 +1,337 @@
+
+There are many ways to detect binary files and choices to be made for serving them.
+
+In the interest of providing gem contents for the purpose of browsing on the web, we think it is best not to serve the binary portions of a gem.
+Binary files are not able to be browsed or diff'd in any meaningful way on a web page.
+
+Avoiding binaries solves many of the problems listed in drawbacks and reduces storage size significantly.
+For example, the `libv8-node-18.13.0.0-arm64-darwin.gem` we examined is 30MB.
+Uncompressed, the whole gem is about 98MB, while the single binary `.a` file is 97MB.
+Not storing the only binary file in this gem saves 99% of the uncompressed storage space while providing almost all the benefit of offering gem browsing.
+
+A user that wants to examine a binary file in a gem is effectively downloading more than the file size of the gem.
+They would be better served downloading the gem for that purpose.
+
+#### Detecting Binary Files
+
+There are many solutions that can be found for detecting binary files.
+We intend at this time to use `grep` to quickly print a list of files that it thinks are binary, and therefore not text searchable.
+We do this by relying on `grep` to output when a binary file matches.
+
+```
+$ grep -RHm1 '^' libv8-node-18.13.0.0-arm64-darwin.gem/* | grep 'Binary'
+Binary file data/vendor/v8/arm64-darwin/libv8/obj/libv8_monolith.a matches
+```
+
+Alternate solutions involve using `file` (the implementation of which varies between different linux and unix based systems) or libmagic (which requires us to add and compile and additional dependency).
+By using `grep` to separate binaries, we use a tool available on every system that behaves in way that is already familiar to most developers.
+
+We still may need to use `file` or `libmagic` to discover the content type of a file.
+The output from `file` on MacOS says it will always have the word `text` in a text file, but it warns that other implementations may not be consistent.
+Using `grep` for determining the `binary` formatted files ensures that we reliably identify files that we don't intend to store.
+
+#
+
+- Why is this design the best in the space of possible designs?
+- What other designs have been considered and what is the rationale for not choosing them?
+- What is the impact of not doing this?
+
+# Unresolved questions
+
+- Feedback on naming. We can tackle column names in individual PRs, but if there is any structural names we'd like to refine, let's address them here.
+- Should we keep the file manifest of yanked gems?
+  Is a yanked gem intended to be trashed completely or just hidden?
+  One possible downside to keeping the file manifest would be if files were committed that leaked private info just by their name or content type.
+  I'm leaning towards not saving yanked file manifests, but happy to hear comments.
+- Should we try to trigger gem ingestion on-demand when a manifest is requested for a gem that doesn't have one yet?
```

Comment:
I think we should give people a button they have to click to say "yeah I want this please go unpack it for me I understand it might take an hour". 

I think that would avoid the web crawler problem, right? And also a "I'ma just clickin around, checkin things out" person from launching 100 background jobs on stuff they never intend to come back for.